### PR TITLE
Migration to Pydantic v2

### LINKED ISSUE #236: Pydantic 2.0 Migration
As part of the pydantic 2.0 release, a suite of breaking changes were released. We need to investigate where this breaks diffsync.

## Issues already identified

- [ ] [`extra="allow"`](https://docs.pydantic.dev/latest/migration/#changes-to-dataclasses) is no longer possible
- [ ] The `__fields__` attribute is deprecated, furthermore it and its replacement `model_fields` no longer work with `__init_subclass__` for us, because their fields are not populated until after this hook runs (see [here](https://github.com/pydantic/pydantic/discussions/7177) for the accompanying discussion)

For more see [here](https://docs.pydantic.dev/latest/migration/).

### PR DESCRIPTION
**Pull request by Kircheneer**
Closes #236

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,16 @@
 # Changelog
```
Comment (Kircheneer) [inline CHANGELOG.md:1]: I've aligned this file to the standard, that's why there are a couple of seemingly unrelated changes

**Review by chadell — APPROVED** [review]
LGTM
